### PR TITLE
store: copy old value when refresh + cas

### LINKED ISSUE #5651: PUT with refresh=true and prevValue sets value to empty on 2.3.6
On 

```
2016-06-13 10:47:54.857444 I | etcdmain: etcd Version: 2.3.6
2016-06-13 10:47:54.857484 I | etcdmain: Git SHA: 128344c
2016-06-13 10:47:54.857487 I | etcdmain: Go Version: go1.6.2
2016-06-13 10:47:54.857489 I | etcdmain: Go OS/Arch: linux/amd64
```

If you do a request:

```"PUT /v2/keys/election?ttl=60&prevValue=2492&refresh=true HTTP/1.1" 200 307```

It sets the value to empty, while keeping the entry

```"PUT /v2/keys/election?ttl=60&prevExit=true&refresh=true HTTP/1.1" 200 307```

Keeps the value.

Expected behaviour:
 * the `value` should be kept if `prevValue` is provided

### PR DESCRIPTION
**Pull request by xiang90**
Fix #5651 

/cc @heyitsanthony @mwitkow

### CODE REVIEW

File: store/store.go [hunk 1]
```diff
@@ -298,6 +298,10 @@ func (s *store) CompareAndSwap(nodePath string, prevValue string, prevIndex uint
 		return nil, err
 	}
 
+	if expireOpts.Refresh {
```
Comment (heyitsanthony) [inline store/store.go]: should this return an error if value != "" in case someone expects this to use the given `value` like a refreshless CAS?

**Comment by heyitsanthony** [comment]
lgtm

**Comment by LLParse** [comment]
Hi @xiang90, I see this is labelled for backport. Which versions is this fix targeted for?

**Comment by heyitsanthony** [comment]
@llparse this was backported into 2.3.7.

**Comment by LLParse** [comment]
@heyitsanthony Thanks...

For reference, I downloaded `https://github.com/coreos/etcd/releases/download/v2.3.7/etcd-v2.3.7-darwin-amd64.zip` on Sep  2 17:39 MST and it exhibited this behavior. I will build the tagged release from source for now.

**Comment by heyitsanthony** [comment]
@LLParse I fetched `2.3.7-darwin-amd64.zip` and I'm seeing `Git SHA: fd17c91` when I run `etcd`. Checking out the release-2.3 repo at that commit, I see 9ce0e8bf8110c9e6045a7f1e7272fee899817be0 right before it. The fix should be built into the binary. How are you testing this?

**Comment by LLParse** [comment]
My code must be the problem. Sorry. It occurs to me that my initial write could've been erroneous when I ran into this issue and I simply ran into this issue when looking for an answer. Alternatively, I'm using the API incorrectly.

This is the initial registration call.
```go
const (
	nodeLockTTL = 30 * time.Second
)

	n.Id = uuid.NewV4().String()
	n.Info = &NodeInfo{n.Config.Hostname}
	nodeInfoJson, _ := json.Marshal(n.Info)
	n.InfoJson = string(nodeInfoJson)
	log.Printf("Creating node %s\n", n.Id)

	_, err := n.Keys.Set(
		context.Background(),
		fmt.Sprintf("/%s/nodes/%s", n.Config.ClusterName, n.Id),
		n.InfoJson,
		&etcd.SetOptions{TTL: nodeLockTTL})
```

This is the call that I perceived as setting the value to an empty string.
```go
			_, err := n.Keys.Set(
				context.Background(),
				fmt.Sprintf("/%s/nodes/%s", n.Config.ClusterName, n.Id),
				"",
				&etcd.SetOptions{TTL: nodeLockTTL, Refresh: true})
```

**Comment by LLParse** [comment]
Whoooops, definitely my mistake. It is functioning properly for me now - I must've messed up the initial set. My apologies, please disregard this.

**Pull request merged**